**STUDENT: Joel S. Mollel**

**NUMBER: C00313599**

**ALGORITHM: Support Vector Machines (SVM)**

**CHANGELOG**

This is own code with assistance from the references mentioned. In this SVM implementation:

1. Trained the SVM model using the drug200.csv dataset

2. Tested the model accuracy

3. Created a Medical drug prescription System

The overall weight of this work is 68% due to the simplicity of the model implementation code. The project took me more than four hours in total due to the fact that I was learning it while implementing.

Using another dataset and perform other operations, and simulate as an app

I will use the dataset drug200.csv, our drug recommendation dataset. A drug to be prescribed to patient is predicted using parameters like age, weight, sex, BP, cholesterol, and Na-to-K.

*Step1: Import the libraries and the dataset*

In [1]:
import numpy as np
import pandas as pd
from sklearn.svm import SVC
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

df = pd.read_csv('drug200.csv')

*Step2: Encoding categorical features and target variable (Drug)*

In [2]:
label_encoders = {}
for column in ['Sex', 'BP', 'Cholesterol']:  # features ('Sex', 'BP', 'Cholesterol')
    le = LabelEncoder()
    df[column] = le.fit_transform(df[column])
    label_encoders[column] = le

le_target = LabelEncoder()
df['Drug'] = le_target.fit_transform(df['Drug'])  # target variable (Drug)

*Step3: Defining features and target variable/label*

In [4]:
X = df[['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K']].values 
y = df['Drug'].values 

*Step4: Splitting data into training (70%) and testing (30%) datasets*

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

kernels = ['linear', 'poly', 'rbf', 'sigmoid']  # different kernels

*Step5: Training the model and testing accuracy*

In [6]:
for kernel in kernels:
    model = SVC(kernel=kernel) 
    model.fit(X_train, y_train)
    
    # Testing accuracy
    accuracy = model.score(X_test, y_test)
    print(f'Accuracy for {kernel} kernel: {accuracy:.4f}')

Accuracy for linear kernel: 0.9833
Accuracy for poly kernel: 0.6500
Accuracy for rbf kernel: 0.7333
Accuracy for sigmoid kernel: 0.4000


**SVM Medical Drugs Prescription System**

The implementation of SVM Drugs Prescription System uses same code as above in addition to lines that will be prompting user input.

**Steps**

*Step1: Importing libraries and dataset*

In [7]:
import pandas as pd
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

data = pd.read_csv('drug200.csv')

*Step2: Processing the data, declaration of features and target*

In [8]:
X = data[['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K']]  # Features
y = data['Drug']  # Target

*Step3: Encoding the categorical columns*

In [9]:
label_encoder_sex = LabelEncoder()
X.loc[:, 'Sex'] = label_encoder_sex.fit_transform(X['Sex'])  # make F = 0, M = 1

label_encoder_bp = LabelEncoder()
X.loc[:, 'BP'] = label_encoder_bp.fit_transform(X['BP'])  # make LOW = 0, NORMAL = 1, HIGH = 2

label_encoder_cholesterol = LabelEncoder()
X.loc[:, 'Cholesterol'] = label_encoder_cholesterol.fit_transform(X['Cholesterol'])  # make NORMAL = 0, HIGH = 1

*Step4: Splitting the dataset, 70% training and 30% testing datasets*

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

*Step5: Train the model and prompt user input, encode the inputs and predict the drug*

We will use Linear kernel because it is the one that had best performance in our model test with the accuracy of 0.9833

In [11]:
model = SVC(kernel='linear', C=1.0)
model.fit(X_train, y_train)

print("Enter the following details to predict the drug type:")

age = float(input("Age: "))
sex = input("Sex (M/F): ").strip().upper() 
bp = input("Blood Pressure (LOW, NORMAL, HIGH): ").strip().upper()
cholesterol = input("Cholesterol (NORMAL, HIGH): ").strip().upper()
na_to_k = float(input("Sodium to Potassium Ratio (Na_to_K): "))

if sex not in ['M', 'F']:
    print("Invalid input for sex! Please enter M or F.")
    sex = 'M' 
if bp not in ['LOW', 'NORMAL', 'HIGH']:
    print("Invalid input for Blood Pressure! Defaulting to NORMAL.")
    bp = 'NORMAL'
if cholesterol not in ['NORMAL', 'HIGH']:
    print("Invalid input for Cholesterol! Defaulting to NORMAL.")
    cholesterol = 'NORMAL'
    
sex = label_encoder_sex.transform([sex])[0]   ##transformation starts
bp = label_encoder_bp.transform([bp])[0]
cholesterol = label_encoder_cholesterol.transform([cholesterol])[0]

input_data = pd.DataFrame([[age, sex, bp, cholesterol, na_to_k]], columns=['Age', 'Sex', 'BP', 'Cholesterol', 'Na_to_K'])

predicted_drug = model.predict(input_data)

print(f"Predicted drug: {predicted_drug[0]}")

Enter the following details to predict the drug type:


Age:  32
Sex (M/F):  m
Blood Pressure (LOW, NORMAL, HIGH):  low
Cholesterol (NORMAL, HIGH):  normal
Sodium to Potassium Ratio (Na_to_K):  22


Predicted drug: drugY


***REFERENCES***

Torabi, T., 2022. Drug Classification Dataset. [online] Available at: https://www.kaggle.com/datasets/taranomtorabi/drug200 [Accessed 20 February 2025].

Support Vector Machines. (2025). YouTube video, added by Simplilearn [online]. Available at: https://www.youtube.com/watch?v=_PwhiWxHK8o&pp=ygUXc3VwcG9ydCB2ZWN0b3IgbWFjaGluZXM%3D [Accessed 10 Feb. 2025].

Stanford Online. (2022). Support Vector Machine (SVM) Explained - What Is The SVM Algorithm?. YouTube video, added by Stanford Online [online]. Available at: https://www.youtube.com/watch?v=lDwow4aOrtg [Accessed 05 Feb. 2025].